pkg/trace/api/otlp.go

```diff
@@ -94,6 +115,23 @@ func (o *OTLPReceiver) Stop() {
 // Export implements ptraceotlp.Server
 func (o *OTLPReceiver) Export(ctx context.Context, in ptraceotlp.ExportRequest) (ptraceotlp.ExportResponse, error) {
 	defer timing.Since("datadog.trace_agent.otlp.process_grpc_request_ms", time.Now())
+
+	select {
+	// Wait for the semaphore to become available, allowing the handler to
+	// decode its payload.
+	// Afer the configured timeout, respond without ingesting the payload,
+	// returning an error.
+	case o.recvsem <- struct{}{}:
+	case <-time.After(time.Duration(o.conf.DecoderTimeout) * time.Millisecond):
+		// this payload can not be accepted
+		return ptraceotlp.NewExportResponse(), fmt.Errorf("Too Many Requests.")
+	}
+	defer func() {
```

Comment:
Would this make more sense in the select/case statement of adding to the channel? Or would that be wrong?

Reply:
The goal here is to block the threads before they decode the messages, to reduce CPU and memory that is devoted to decoding and storing payloads when they can't be enqueued to the processor right away.

I'm not sure this is the right place for it, but it should go as far forward in the client-server communication as possible (ideally before we even read in the message from the client). Because it's gRPC I'll need to look closer about how to do that.